vpr/src/place/analytic_placer.h

```diff
@@ -0,0 +1,249 @@
+ * returns the ID of the head block
+ */
+ClusterBlockId macro_head(ClusterBlockId blk);
+
+class AnalyticPlacer {
+  public:
+    /*
+     * @brief Constructor of AnalyticPlacer, currently initializes AnalyticPlacerCfg for the analytic placer
+     * To tune these parameters, change directly in constructor
+     */
+    AnalyticPlacer();
+
+    /*
+     * @brief top function of analytic placement
+     * Takes the random initial placement from place.cpp through g_vpr_ctx
+     * Repeat the following until stopping criteria is met:
+     * 	* Formulate and solve equations in x, y directions for 1 type of logial block
+     * 	* Instantiate CutSpreader to spread and strict_legalize
+     *
+     * The final legal placement is passed back to annealer in g_vpr_ctx.mutable_placement()
+     */
+    void ap_place();
+
+  private:
+    // for CutSpreader to access legal_pos, block_locs, etc.
+    friend class CutSpreader;
+
+    // AP parameters that can influence it's behavior
+    struct AnalyticPlacerCfg {
+        float alpha;                        // anchoring strength of pseudo-connections
+        float beta;                         // over-utilization factor
+        int criticalityExponent;            // not currently used, @see build_equations()
+        int timingWeight;                   // not currently used, @see build_equations()
+        float solverTolerance;              // parameter of the solver
+        int spread_scale_x, spread_scale_y; // see CutSpreader::expand_regions()
+    };
+
+    AnalyticPlacerCfg tmpCfg; // TODO: PlacerHeapCfg should be externally configured & supplied
+
+    // @see initial_placement.cpp load_legal_locations()
+    std::vector<std::vector<int>> num_legal_pos;
+    std::vector<std::vector<std::vector<t_pl_loc>>> legal_pos;
+
+    // marker for blks not solved in current iteration
+    int32_t dont_solve = std::numeric_limits<int32_t>::max();
+
+    // row number in the system of linear equations for each block
+    std::unordered_map<ClusterBlockId, int32_t> blk_row_num;
+
+    // Encapsulates 3 types of locations for each logic block
+    struct BlockLocation {
+        t_pl_loc loc;       // real, up-to-date location of the logic block
+        t_pl_loc legal_loc; // legalized location, used to create psudo connections
+        double rawx, rawy;  // raw location storing float result from matrix solver
+    };
+
+    // Lookup from blockID to block location
+    std::unordered_map<ClusterBlockId, BlockLocation> blk_locs;
+
+    // The set of blks actually placed.
+    // Excludes children of macros.
+    std::vector<ClusterBlockId> place_blks;
+
+    // blocks being solved in the current equation
+    // subset of all blocks being placed
+    std::vector<ClusterBlockId> solve_blks;
+
+    /*
+     * number of blocks on each tile
+     * used to print a simple placement graph in print_place
+     */
+    std::vector<std::vector<int>> overlap;
```

Comment:
Shouldn't be a class member variable; can reduce scope to a local variable that is returned by overlap() and printed by print_overlap(overlap&).